[clang-tidy] Fix assert in modernize-use-std-format/print

mikecrowe opened the pull request:
Ensure that FormatStringConverter's constructor fails with a sensible error message rather than asserting if the format string is not a narrow string literal.

Also, ensure that we don't even get that far in modernize-use-std-print and modernize-use-std-format by checking that the format string parameter is a char pointer.

Fixes #92896

mikecrowe committed:
[clang-tidy] Fix assert in modernize-use-std-format/print

Ensure that FormatStringConverter's constructor fails with a sensible
error message rather than asserting if the format string is not a narrow
string literal.

Also, ensure that we don't even get that far in modernize-use-std-print
and modernize-use-std-format by checking that the format string
parameter is a char pointer.

Fixes #92896

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:
```diff
@@ -20,6 +20,11 @@ namespace clang::tidy::modernize {
 AST_MATCHER(StringLiteral, isOrdinary) { return Node.isOrdinary(); }
+AST_MATCHER(QualType, isSimpleChar) {
+  const auto ActualType = Node.getTypePtr();
+  return ActualType->isSpecificBuiltinType(BuiltinType::Char_S) ||
```
Note: you ay also verify that ActualType is not null

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/UseStdPrintCheck.cpp:
```diff
@@ -20,6 +20,11 @@ namespace clang::tidy::modernize {
 
 namespace {
 AST_MATCHER(StringLiteral, isOrdinary) { return Node.isOrdinary(); }
+AST_MATCHER(QualType, isSimpleChar) {
+  const auto ActualType = Node.getTypePtr();
+  return ActualType->isSpecificBuiltinType(BuiltinType::Char_S) ||
+         ActualType->isSpecificBuiltinType(BuiltinType::Char_U);
+}
```
consider moving this to some exist file in utils directory

mikecrowe replied:
OK. I probably ought to do the same with `isOrdinary` too, but I'll do that in a separate change.

PiotrZSL reviewed (approved):
Overall LGTM, consider adding release note entry for modernize-use-std-print (as it were added in previous release)

mikecrowe committed:
fixup! Improve isSimpleChar and move it into utils/Matchers.h

mikecrowe committed:
fixup! Mention fix in release notes

Don't mention modernize-use-std-format check since it is new in this
version anyway.

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -351,6 +351,10 @@ Changes in existing checks
   <clang-tidy/checks/modernize/use-starts-ends-with>` check to also handle
   calls to ``compare`` method.
 
+- Improved :doc:`modernize-use-std-format` and
```
remove modernize-use-std-format from here, as it's new check, only leave std-print

mikecrowe replied:
I spotted that just too late...

PiotrZSL reviewed (approved):
LGTM

mikecrowe commented:
> LGTM

Thanks for the review. Please can you squash and land it for me? I don't have permission to do so.

mikecrowe commented:
Thank you @PiotrZSL!
